## Lab 03.01 - Introduction to Pandas and Data Operations
In this lab, you'll learn how to use the pandas library to analyze real-world sales data. By the end of this lab, you'll be comfortable with the fundamental operations in pandas.

### Part 0 - Intro to Pandas
Before we dive into using pandas, let's understand what it is and why it's important in data science.

To answer the questions, edit the markdown cell and put your answer below the question. 

**Make sure to save the markdown cell, by pressing the ✓ (check) icon in the top right after answering the questions**

Research and critically think about the following questions:

##### Question 00
What is Pandas, and what are its primary functions in data visualization for data science?
- **Answer:**
Pandas is a Python library buit on top of the matplotlib library that is used to process data. Pandas provide what matplotlib plot provide addition to some of its own data structures and data analysis tools

##### Question 01
What does "data manipulation" mean in the context of pandas and data science field?
- **Answer:**
"data manipulation" mean in the context of pandas and data science field means to process datat in a way that it becomes easier to read.

##### Question 02
What kind of files can pandas read? List at least 3:
- **Answer:**
1. csv
2. excel
3. sql

#### 0.0 - Hands-On Exploration
Let's start by importing pandas and exploring its basic functionality. Type these commands and write down what you observe:

First, let's install the matplotlib library, in your terminal run the following command:

```bash
pip3 install pandas
```

Once installed, let's start by importing pandas and exploring its basic functionality. 

In [2]:
import pandas as pd

Now lets run create a list, and see what happens when we use it with pandas.

In [3]:
# Create a simple list of numbers
numbers = [1, 2, 3, 4, 5]

# Convert it to a pandas Series
series = pd.Series(numbers)

# Print the result
print(series)

0    1
1    2
2    3
3    4
4    5
dtype: int64


##### Question 03
What did you notice about the output format?
- **Answer:**
I notice that the output is in a 2 column table format with 5 rows corrsponding to the 5 numbers in the list.
##### Question 04
How is this different from a regular Python list? What are the numbers on the left side?
- **Answer:**
This is different from a regular list in that the index is displayed in the left column along side the value of the index in the second column.

### Part 1 - Reading and Exploring Data

#### 01.00 - Loading Data
Let's start by loading our sales data that you downloaded alongside this lab. In pandas, we can read various file formats, but CSV (Comma-Separated Values) is one of the most common.

In [4]:
sales_df = pd.read_csv('sales_data.csv')

When using pandas to load data, in ingests it into what is known as a "dataframe", this allows us to do advanced manipulation.

#### 01.01 - Viewing Data Samples
Let's explore the fundamental methods for understanding our data structure and content. We'll look at each method individually and understand what it tells us about our data.

In [7]:
print(sales_df.head(20))

          Date         Product         Category    Price  Units    Total  \
0   2023-01-01        Notebook  Office Supplies    18.33      2    35.56   
1   2023-01-01          Laptop      Electronics  1943.35      4  7773.40   
2   2023-01-01          Tablet      Electronics   436.05      2   872.10   
3   2023-01-01       USB Drive      Accessories   116.48      3   349.44   
4   2023-01-01         Pen Set  Office Supplies    42.14      4   168.56   
5   2023-01-01         Pen Set  Office Supplies    26.14      4    90.97   
6   2023-01-01      Laptop Bag      Accessories    94.14      3   245.71   
7   2023-01-01         Pen Set  Office Supplies    46.51      3   139.53   
8   2023-01-01            Lamp        Furniture   353.09      5  1518.29   
9   2023-01-01      Smartwatch      Electronics   893.71      3  2681.13   
10  2023-01-01          Laptop      Electronics   802.12      3  2093.53   
11  2023-01-01   Monitor Stand      Accessories    71.68      5   358.40   
12  2023-01-

In [6]:
print(sales_df.tail())

            Date       Product         Category    Price  Units    Total  \
3766  2023-12-31    Smartwatch      Electronics  1730.72      2  3461.44   
3767  2023-12-31          Desk        Furniture   462.57      1   462.57   
3768  2023-12-31        Laptop      Electronics   560.79      2  1121.58   
3769  2023-12-31  Office Chair        Furniture   116.04      2   232.08   
3770  2023-12-31      Notebook  Office Supplies    18.15      4    72.60   

      Discount Region        Sales_Rep Payment_Method Customer_Segment  
3766       0.0  South  Michael Johnson     Debit Card       Individual  
3767       0.0  North       John Smith     Debit Card       Government  
3768       0.0   East      David Brown     Debit Card   Small Business  
3769       0.0  North       Emma Davis    Credit Card        Corporate  
3770       0.0  North       Emma Davis         PayPal       Individual  


##### Question 04
What is the difference between the `tail()` and `head()` commands? What would be use of either command be?
- **Answer:**
`tail()` prints the data of the first 5 samples and can be used to view the first 5 data samples in a data set. `head()` prints the data of the last 5 samples and can be used to view the last 5 data samples in a data set. Both are ordered from top to bottom.

##### Question 05
What happens when you put a number in the `head()` function? What changed?
- **Answer:**
When you put a number in the `head()` function the head function will print a amount of data samples equal to the number place in the ().

#### 01.02 - Understanding DataFrame Structure
Let's examine the basic properties of our DataFrame:

In [8]:
print(sales_df.columns)

Index(['Date', 'Product', 'Category', 'Price', 'Units', 'Total', 'Discount',
       'Region', 'Sales_Rep', 'Payment_Method', 'Customer_Segment'],
      dtype='object')


In [9]:
print(sales_df.dtypes)

Date                 object
Product              object
Category             object
Price               float64
Units                 int64
Total               float64
Discount            float64
Region               object
Sales_Rep            object
Payment_Method       object
Customer_Segment     object
dtype: object


##### Question 06
Compare and contrast the outputs of the `columns` and `dtypes` properties. What is similar what is different?
- **Answer:**
Both `columns` and `dtypes` outs information on the columns, but`columns`  outputs a list name of the columns, and `dtypes` out put the name of the colums and the data type they store.

##### Question 07
What data type is the 'Date' column? Is this what you expected? 
- **Answer:**
The 'Date' column has a data type of object, which was unexpected as i expected it to be a string.

##### Question 08
Why might pandas choose different data types for different columns?
- **Answer:**
pandas might choose different data types for different columns because some columns contain information that needs to be used in math while other do not and are not numbers.

#### 01.03 - Data Information Summary
The `info()` method provides a concise summary of our DataFrame:

In [10]:
print(sales_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3771 entries, 0 to 3770
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date              3771 non-null   object 
 1   Product           3771 non-null   object 
 2   Category          3771 non-null   object 
 3   Price             3771 non-null   float64
 4   Units             3771 non-null   int64  
 5   Total             3771 non-null   float64
 6   Discount          3771 non-null   float64
 7   Region            3771 non-null   object 
 8   Sales_Rep         3693 non-null   object 
 9   Payment_Method    3739 non-null   object 
 10  Customer_Segment  3771 non-null   object 
dtypes: float64(3), int64(1), object(7)
memory usage: 324.2+ KB
None


##### Question 09
What information does `info()` tell us, how could this information be useful?
- **Answer:**
`info()` tell us the name of each column, the Non-Null Count of each column, data type of each column, number of each data type used by the column, total memory usage. This information can be useful when determining the data type used used by the data set, the memory used by the data, and whether or not data is missing.

##### Question 10
How much memory is our DataFrame using? 
- **Answer:**
The DataFrame is using 324.2+ KB worth of memory.

##### Question 11
What does "null" mean? What impact could having "null" values in a dataset?
- **Answer:**
null is a constant that repersent a lack of a value and in a dataset it could repersent a lack of data gather on the related subject.

#### 01.04 - Numerical Summaries
The `describe()` method provides statistical summaries for numerical columns:

In [11]:
print(sales_df.describe())

             Price        Units         Total     Discount
count  3771.000000  3771.000000   3771.000000  3771.000000
mean    397.093911     2.985150   1159.701854     0.021594
std     514.524434     1.410003   1750.424109     0.041496
min       4.810000     1.000000      4.990000     0.000000
25%      40.165000     2.000000    110.730000     0.000000
50%     134.860000     3.000000    372.870000     0.000000
75%     516.490000     4.000000   1472.525000     0.020000
max    2073.680000     5.000000  10151.100000     0.150000


##### Question 12
What statistics does the `describe()` function give us?
- **Answer:**
`describe()` give the statistics of count, mean, std, min, 25%, 50%, 75%, and max.

##### Question 13
Which columns did the `describe()` function perform statistics on? Why? (Hint: why didnt we see 'Product' in the output?)
- **Answer:**
The `describe()` function perform statistics on columns 3-6 of the Price, nits, Total, and Discount column. This is the case because the Price, nits, Total, and Discount column were the only ones to store number(interger or float).

In [15]:
print(sales_df[['Price', 'Units']].describe())

             Price        Units
count  3771.000000  3771.000000
mean    397.093911     2.985150
std     514.524434     1.410003
min       4.810000     1.000000
25%      40.165000     2.000000
50%     134.860000     3.000000
75%     516.490000     4.000000
max    2073.680000     5.000000


##### Question 14
What is different in the codeblock above compared to the previous one. How did this affect the output?
- **Answer:**
The code block above compared to the previous one has a list of 2 column names inside brackets connected to `sales_df`, this caused the code to only output the count, mean, std, min, 25%, 50%, 75%, and max of the two listed columns, which arePrice and Units.

In [16]:
print(sales_df['Category'].unique())

['Office Supplies' 'Electronics' 'Accessories' 'Furniture']


##### Question 15
How many unique product categories are in our data set?
- **Answer:**
There are 4 unique product categories in our data set.

In [ ]:
print("Output 1:")
print(sales_df['Category'].value_counts())

print('')
print('')

print("Output 2:")
print(sales_df['Category'].value_counts(normalize=True))

##### Question 16
What is the difference between the two outputs?
- **Answer:**

##### Question 17
What is our most common product category?
- **Answer:**

##### Excercise 00
Using what you've learned, create a code cell direclty below this one to answer these questions:

- What is the date range of our sales data? (Hint: try `min()` and `max()` functions)
- How many unique products do we sell?
- What is our most common payment method?

### Part 2 - Data Selection and Filtering

#### 02.01 - Column Selection
There are multiple ways to select columns in pandas. Let's explore them:

In [ ]:
# Select single column
prices = sales_df['Price']

# Select multiple columns
product_info = sales_df[['Product', 'Price', 'Units']]

##### Question 17
Compare and contrast how we select multiple columns vs a single column.
- **Answer:**

##### Excercise 01
Using what you've learned, create a code cell direclty below this one to select the 'Region' and 'Category' columns and store them in a variable called 'location_data'.

#### 02.01 - Filtering Data
Let's learn how to filter our data based on conditions:

In [ ]:
expensive_items = sales_df[sales_df['Price'] > 500]

north_sales = sales_df[sales_df['Region'] == 'North']

print("Expensive items:")
print(expensive_items)
print("\nNorth region sales:")
print(north_sales)

##### Question 18
What is `sales_df['Price'] > 500` accomplishing in our code? How might we use this in data science workflows?
- **Answer:**

##### Question 19
What logic operator would we use to combine multiple conditonals when filtering columns?
- **Answer:**

##### Excercise 02
Using what you've learned, create a code cell direclty below this one to create the following filters:
- The number of units sold was greater than 5
- The product category is 'Electronics'

#### Part 3 - Final Summary
Use your experience completing this lab, and the code block below to answer the following summary questions

##### Question 20
In your own words, explain what pandas is and why it's useful for data analysis:
- **Answer:**

##### Question 21
How many unique products are in the dataset?
- **Answer:**

##### Question 22
What was the most popular product category in the `North` region? 
- **Answer:**